<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0

  Attempting uninstall: yfinance
    Found existing installation: yfinance 0.2.4
    Uninstalling yfinance-0.2.4:
      Successfully uninstalled yfinance-0.2.4


'mamba' is not recognized as an internal or external command,
operable program or batch file.


     ---------------------------------------- 0.0/153.3 kB ? eta -:--:--
     ---------------------------------------- 0.0/153.3 kB ? eta -:--:--
     ------------------ -------------------- 71.7/153.3 kB 4.1 MB/s eta 0:00:01
     ----------------------- --------------- 92.2/153.3 kB 1.7 MB/s eta 0:00:01
     ------------------------------ ------- 122.9/153.3 kB 1.0 MB/s eta 0:00:01
     ----------------------------------- -- 143.4/153.3 kB 1.1 MB/s eta 0:00:01
     ------------------------------------ 153.3/153.3 kB 831.7 kB/s eta 0:00:00
  Attempting uninstall: nbformat
    Found existing installation: nbformat 5.7.0
    Uninstalling nbformat-5.7.0:
      Successfully uninstalled nbformat-5.7.0


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
conda-repo-cli 1.0.41 requires requests_mock, which is not installed.
anaconda-client 1.12.0 requires nbformat>=4.4.0, but you have nbformat 4.2.0 which is incompatible.
conda-repo-cli 1.0.41 requires clyent==1.2.1, but you have clyent 1.2.2 which is incompatible.
conda-repo-cli 1.0.41 requires nbformat==5.4.0, but you have nbformat 4.2.0 which is incompatible.
conda-repo-cli 1.0.41 requires requests==2.28.1, but you have requests 2.31.0 which is incompatible.
jupyter-server 1.23.4 requires nbformat>=5.2.0, but you have nbformat 4.2.0 which is incompatible.
nbclient 0.5.13 requires nbformat>=5.0, but you have nbformat 4.2.0 which is incompatible.
nbconvert 6.5.4 requires nbformat>=5.1, but you have nbformat 4.2.0 which is incompatible.


In [45]:
import yfinance as yf

import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots


In [29]:
!pip install -U yfinance pandas

  Obtaining dependency information for yfinance from https://files.pythonhosted.org/packages/e6/b3/388ab967a387cc92926f70e97688dd9a7189b29a0773db815ffc5289e2b5/yfinance-0.2.31-py2.py3-none-any.whl.metadata
  Using cached yfinance-0.2.31-py2.py3-none-any.whl.metadata (11 kB)
  Obtaining dependency information for pandas from https://files.pythonhosted.org/packages/97/d8/dc2f6bff06a799a5603c414afc6de39c6351fe34892d50b6a077df3be6ac/pandas-2.1.3-cp311-cp311-win_amd64.whl.metadata
     ---------------------------------------- 0.0/341.8 kB ? eta -:--:--
     ---- -------------------------------- 41.0/341.8 kB 653.6 kB/s eta 0:00:01
     ---------- ---------------------------- 92.2/341.8 kB 1.3 MB/s eta 0:00:01
     --------------- -------------------- 143.4/341.8 kB 944.1 kB/s eta 0:00:01
     ------------------ ----------------- 174.1/341.8 kB 952.6 kB/s eta 0:00:01
     ------------------ ----------------- 174.1/341.8 kB 952.6 kB/s eta 0:00:01
     ------------------ ----------------- 174.

In [46]:
!pip install pandas==1.2.2

     ---------------------------------------- 0.0/5.5 MB ? eta -:--:--
     ---------------------------------------- 0.0/5.5 MB ? eta -:--:--
     ---------------------------------------- 0.0/5.5 MB ? eta -:--:--
     ---------------------------------------- 0.0/5.5 MB ? eta -:--:--
     ---------------------------------------- 0.1/5.5 MB 544.7 kB/s eta 0:00:10
     ---------------------------------------- 0.1/5.5 MB 544.7 kB/s eta 0:00:10
     ---------------------------------------- 0.1/5.5 MB 544.7 kB/s eta 0:00:10
     ---------------------------------------- 0.1/5.5 MB 544.7 kB/s eta 0:00:10
      --------------------------------------- 0.1/5.5 MB 310.3 kB/s eta 0:00:18
      --------------------------------------- 0.1/5.5 MB 327.2 kB/s eta 0:00:17
      --------------------------------------- 0.1/5.5 MB 300.4 kB/s eta 0:00:18
      --------------------------------------- 0.1/5.5 MB 300.4 kB/s eta 0:00:18
     - -------------------------------------- 0.2/5.5 MB 305.7 kB/s eta 0:00

  error: subprocess-exited-with-error
  
  exit code: 1
  
  [1261 lines of output]
  <string>:44: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-cpython-311
  creating build\lib.win-amd64-cpython-311\pandas
  copying pandas\conftest.py -> build\lib.win-amd64-cpython-311\pandas
  copying pandas\testing.py -> build\lib.win-amd64-cpython-311\pandas
  copying pandas\_testing.py -> build\lib.win-amd64-cpython-311\pandas
  copying pandas\_typing.py -> build\lib.win-amd64-cpython-311\pandas
  copying pandas\_version.py -> build\lib.win-amd64-cpython-311\pandas
  copying pandas\__init__.py -> build\lib.win-amd64-cpython-311\pandas
  creating build\lib.win-amd64-cpython-311\pandas\api
  copying pandas\api\__init__.py -> build\lib.win-amd64-cpython-311\pandas\api
  creating build\lib.win-amd64-cpython-311\pandas\arrays
  copying pandas\arrays\__i

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [2]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [47]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [43]:
tesla=yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [49]:
tesla_data = tesla.history(period='max')


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [ ]:
tesla_data.reset_index(inplace=True)

In [ ]:
tesla_data.head(8)

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [50]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
response=requests.get(url)
if response.status_code==200:
    html_data=response.text
    print("webpage downloaded")
else:
    print("download failed")

webpage downloaded


Parse the html data using `beautiful_soup`.


In [51]:
from bs4 import BeautifulSoup
soup=BeautifulSoup(html_data,'html.parser')
print("Title",soup.title)
div_element=soup.find('div')
print("first element of div",div_element)

Title <title>Tesla Revenue 2010-2022 | TSLA | MacroTrends</title>
first element of div <div class="header_content_container container-fluid">
<div class="header_parent_container">
<div class="header_container">
<div class="header_logo col-xs-2">
<a class="logo" href="https://www.macrotrends.net" title="MacroTrends Home Page"><img src="/assets/images/logo_bright1.png"/></a>
</div>
<div class="col-xs-1 pull-right" style="padding-top:8px; margin-right:10px; margin-left:0px; padding-left:0px;">
<div id="ic_88x31_1">
</div>
</div>
<div class="col-xs-5 pull-right" style="padding-top:8px;">
<form>
<div class="typeahead__container">
<div class="typeahead__field">
<span class="typeahead__query">
<input autocomplete="off" autofocus="" class="js-typeahead" name="q" placeholder="Search over 200,000 charts..." type="search"/>
</span>
<span class="typeahead__button">
<button type="submit">
<span class="typeahead__search-icon"></span>
</button>
</span>
</div>
</div>
</form>
</div>
</div>
</div>
<div 

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [115]:
soup.find_all("table")

[<table class="historical_data_table table">
 <thead>
 <tr>
 <th colspan="2" style="text-align:center">Tesla Annual Revenue<br/><span style="font-size:14px;">(Millions of US $)</span></th>
 </tr>
 </thead>
 <tbody>
 <tr>
 <td style="text-align:center">2021</td>
 <td style="text-align:center">$53,823</td>
 </tr>
 <tr>
 <td style="text-align:center">2020</td>
 <td style="text-align:center">$31,536</td>
 </tr>
 <tr>
 <td style="text-align:center">2019</td>
 <td style="text-align:center">$24,578</td>
 </tr>
 <tr>
 <td style="text-align:center">2018</td>
 <td style="text-align:center">$21,461</td>
 </tr>
 <tr>
 <td style="text-align:center">2017</td>
 <td style="text-align:center">$11,759</td>
 </tr>
 <tr>
 <td style="text-align:center">2016</td>
 <td style="text-align:center">$7,000</td>
 </tr>
 <tr>
 <td style="text-align:center">2015</td>
 <td style="text-align:center">$4,046</td>
 </tr>
 <tr>
 <td style="text-align:center">2014</td>
 <td style="text-align:center">$3,198</td>
 </tr>
 <tr

In [54]:
if response.status_code == 200:
    # Parse the HTML data with BeautifulSoup
    soup = BeautifulSoup(response.text, 'html.parser')

    # Find the table with Tesla Revenue using its class (replace 'your-table-class' with the actual class)
    table = soup.find('table', {'class': 'historical_data_table table'})

    # Extract data from the table using pandas' read_html
    tables = pd.read_html(str(table))
    

    # Assuming the table you want is the first one, you can modify accordingly if there are multiple tables
    tesla_revenue = tables[0]

    # Display the DataFrame
    print(tesla_revenue)
else:
    print("failed")

    Tesla Annual Revenue (Millions of US $)  \
0                                      2021   
1                                      2020   
2                                      2019   
3                                      2018   
4                                      2017   
5                                      2016   
6                                      2015   
7                                      2014   
8                                      2013   
9                                      2012   
10                                     2011   
11                                     2010   
12                                     2009   

   Tesla Annual Revenue (Millions of US $).1  
0                                    $53,823  
1                                    $31,536  
2                                    $24,578  
3                                    $21,461  
4                                    $11,759  
5                                     $7,000  
6           

In [55]:
tesla_revenue.rename(columns={'Tesla Annual Revenue (Millions of US $)':'Date'},inplace=True)
tesla_revenue.head(2)

Date Tesla Annual Revenue (Millions of US $).1
0  2021                                   $53,823
1  2020                                   $31,536

In [56]:
tesla_revenue.head(8)

Date Tesla Annual Revenue (Millions of US $).1
0  2021                                   $53,823
1  2020                                   $31,536
2  2019                                   $24,578
3  2018                                   $21,461
4  2017                                   $11,759
5  2016                                    $7,000
6  2015                                    $4,046
7  2014                                    $3,198

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [57]:
tesla_revenue["Tesla Annual Revenue (Millions of US $).1"] = tesla_revenue['Tesla Annual Revenue (Millions of US $).1'].str.replace(',|\$',"")

In [58]:
tesla_revenue.head(8)

Date Tesla Annual Revenue (Millions of US $).1
0  2021                                     53823
1  2020                                     31536
2  2019                                     24578
3  2018                                     21461
4  2017                                     11759
5  2016                                      7000
6  2015                                      4046
7  2014                                      3198

Execute the following lines to remove an null or empty strings in the Revenue column.


In [59]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Tesla Annual Revenue (Millions of US $).1'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [60]:
tesla_revenue.tail(5)

Date Tesla Annual Revenue (Millions of US $).1
8   2013                                      2013
9   2012                                       413
10  2011                                       204
11  2010                                       117
12  2009                                       112

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [61]:
data=yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [62]:
game_data=data.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [63]:
game_data.reset_index(inplace=True)
game_data.head(5)

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620128  1.693350  1.603296  1.691666  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658002  1.674834   8389600        0.0   
3 2002-02-19  1.666417  1.666417  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615921  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [64]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
response=requests.get(url)
if response.status_code==200:
    html_data=response.text
    print("success")
else:
    print("failed")

success


Parse the html data using `beautiful_soup`.


In [65]:
from bs4 import BeautifulSoup
soup=BeautifulSoup(html_data,'html.parser')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [66]:
if response.status_code==200:
    soup=BeautifulSoup(response.text,'html.parser')
    table=soup.find('table',{'class':'historical_data_table'})
    tables=pd.read_html(str(table))
    gme_revenue=tables[0]
    print(gme_revenue)

    GameStop Annual Revenue (Millions of US $)  \
0                                         2020   
1                                         2019   
2                                         2018   
3                                         2017   
4                                         2016   
5                                         2015   
6                                         2014   
7                                         2013   
8                                         2012   
9                                         2011   
10                                        2010   
11                                        2009   
12                                        2008   
13                                        2007   
14                                        2006   
15                                        2005   

   GameStop Annual Revenue (Millions of US $).1  
0                                        $6,466  
1                                        $8,285  

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [67]:
gme_revenue.tail(8)

GameStop Annual Revenue (Millions of US $)  \
8                                         2012   
9                                         2011   
10                                        2010   
11                                        2009   
12                                        2008   
13                                        2007   
14                                        2006   
15                                        2005   

   GameStop Annual Revenue (Millions of US $).1  
8                                        $9,551  
9                                        $9,474  
10                                       $9,078  
11                                       $8,806  
12                                       $7,094  
13                                       $5,319  
14                                       $3,092  
15                                       $1,843

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [68]:
make_graph(tesla_data,tesla_revenue,'Tesla')

AttributeError: 'function' object has no attribute 'Date'

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
